Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrectionPaalmanPings.h

```diff
@@ -0,0 +1,126 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2020 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#pragma once
+
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/MatrixWorkspace_fwd.h"
+#include "MantidAlgorithms/DllConfig.h"
+#include "MantidGeometry/Objects/IObject.h"
+#include "MantidKernel/Material.h"
+#include "MantidKernel/V3D.h"
+
+namespace Mantid {
+
+namespace API {
+class Sample;
+}
+namespace Geometry {
+class IDetector;
+class IObject;
+} // namespace Geometry
+
+namespace Algorithms {
+/** AbsorptionCorrectionPaalmanPings : calculate paalman-pings absorption terms
+ */
+/** Expansion of the AbsorptionCorrection algorithm to calculate full
+   paalman-pings absorption terms
+
+*/
+class MANTID_ALGORITHMS_DLL AbsorptionCorrectionPaalmanPings
+    : public API::Algorithm {
+public:
+  /// (Empty) Constructor
+  AbsorptionCorrectionPaalmanPings();
+  /// Algorithm's name
+  const std::string name() const override {
+    return "AbsorptionCorrectionPaalmanPings";
+  }
+  /// Algorithm's category for identification
+  const std::string category() const override {
+    return "CorrectionFunctions\\AbsorptionCorrections";
+  }
+  /// Algorithm's summary
+  const std::string summary() const override {
+    return "Calculates the attenuation due to absorption "
+           "and single scattering in a generic sample shape for all "
+           "Paalmin-pings terms.";
+  }
+
+  const std::vector<std::string> seeAlso() const override {
+    return {"SetSampleMaterial", "AbsorptionCorrection",
+            "ApplyPaalmanPingsCorrection"};
+  }
+
+  /// Algorithm's version
+  int version() const override { return (1); }
+
+private:
+  /// Initialisation code
+  void init() override;
+  std::map<std::string, std::string> validateInputs() override;
+  /// Execution code
+  void exec() override;
+
+  void retrieveBaseProperties();
+  void constructSample(API::Sample &sample);
+  void calculateDistances(const Geometry::IDetector &detector,
+                          std::vector<double> &sample_L2s,
+                          std::vector<double> &sample_container_L2s,
+                          std::vector<double> &container_L2s,
+                          std::vector<double> &container_sample_L2s) const;
+  void
+  doIntegration(double &integral, double &crossIntegral,
```

Comment:
Do you have plans to clean up member functions that currently have VERY large numbers of arguments?

Reply:
No